tooling/autodispose-error-prone-checker/src/main/java/com/uber/autodispose/error/prone/checker/MissingAutoDisposeErrorChecker.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (C) 2017. Uber Technologies
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.autodispose.error.prone.checker;
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.ImmutableList;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.ErrorProneFlags;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.tools.javac.code.Type;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.errorprone.BugPattern.SeverityLevel.ERROR;
+import static com.google.errorprone.matchers.Matchers.instanceMethod;
+
+/**
+ * Checker for subscriptions not binding to lifecycle in components with lifecycle.
+ * Use -XepOpt:AutoDisposeLeakCheck to add support for custom components with lifecycle.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    name = "MissingAutoDisposeErrorChecker",
+    summary = "Always apply an Autodispose scope before subscribing",
+    tags = {BugPattern.StandardTags.CONCURRENCY},
+    severity = ERROR
+)
+public final class MissingAutoDisposeErrorChecker extends BugChecker
+    implements MethodInvocationTreeMatcher {
+
+  private static final String AS = "as";
+  private static final ImmutableList<MethodMatchers.MethodNameMatcher> AS_CALL_MATCHERS;
+  private static final ImmutableList<MethodMatchers.MethodNameMatcher> SUBSCRIBE_MATCHERS;
+  private static final String SUBSCRIBE = "subscribe";
+
+  private final Matcher<MethodInvocationTree> matcher;
+
+  public MissingAutoDisposeErrorChecker() {
+    this(ErrorProneFlags.empty());
+  }
+
+  public MissingAutoDisposeErrorChecker(ErrorProneFlags flags) {
+    Optional<ImmutableList<String>> inputClasses = flags.getList("AutoDisposeLeakCheck");
+    ImmutableList<String> classesWithLifecycle = new ImmutableList.Builder<String>()
```

Comment:
This list seems to still be missing support fragments, `LifecycleOwner`, and `ScopeProvider`. Is that intentional? Also, we should have a way to allow the user to disable these built-in classes if they want.

Reply:
I can add these. Not sure what is the best way to disable it? Another flag?

Reply:
Another option is to treat flag as an override and not an addition. The developers may have to pass all these base classes themselves. Con - The callers need to be aware of all the base cases. Pros - Simple.

Reply:
Went with 2nd as it is much simpler and requirements for different projects can be different. https://github.com/uber/AutoDispose/pull/156/commits/cf07e466c34ba52189a68b0bf97751f5a08dad37